ts2mls/js/src/main/scala/ts2mls/TSNamespace.scala

```diff
@@ -0,0 +1,74 @@
+package ts2mls
+
+import scala.collection.mutable.{HashMap, ListBuffer}
+import types._
+
+class TSNamespace(name: String, parent: Option[TSNamespace]) {
+  private val subSpace = HashMap[String, TSNamespace]()
+  private val members = HashMap[String, TSType]()
+
+  // write down the order of members
+  // easier to check the output one by one
+  private val order = ListBuffer.empty[Either[String, String]]
+
+  def derive(name: String): TSNamespace =
+    if (subSpace.contains(name)) subSpace(name) // if the namespace has appeared in another file, just return it
+    else {
+      val sub = new TSNamespace(name, Some(this))
+      subSpace.put(name, sub)
+      order += Left(name)
+      sub
+    }
+
+  def put(name: String, tp: TSType): Unit =
+    if (!members.contains(name)) {
+      order += Right(name)
+      members.put(name, tp)
+    }
+    else members.update(name, tp)
+
+  def get(name: String): TSType =
+    members.getOrElse(name,
+      if (!parent.isEmpty) parent.get.get(name) else throw new Exception(s"member $name not found."))
+
+  def containsMember(name: String, searchParent: Boolean = true): Boolean =
+    if (parent.isEmpty) members.contains(name) else (members.contains(name) || (searchParent && parent.get.containsMember(name)))
+
+  def generate(writer: JSWriter): Unit =
+    order.toList.foreach((p) => p match {
+      case Left(name) => subSpace(name).generate(writer)
+      case Right(name) => {
+        val mem = members(name)
+        val fullName = getFullPath(name)
+        mem match {
+          case inter: TSIntersectionType => { // overloaded functions
+            val typeParams = TSIntersectionType.getOverloadTypeParameters(inter).map((t) => t.name)
```

Comment:
The logic of `getOverloadTypeParameters` seems very wrong. So if `foo: (<A>() -> A) & (<B>() -> B)` you'll get `List(A, B)` as the result...!

Reply:
seems code like this is also acceptable:
```ts
function baz<T>(): T;
function baz<U, T>(): U;
function baz<U, T, P>(): P;
function baz() {
  // do something here
}
```

any better idea to deal with it? :worried:

Reply:
> seems code like this is also acceptable

Of course; that's the point!

You should NEVER EVER return the wrong result out of convenience when you don't know what to do. I can't emphasize this enough. Correctness comes first, otherwise there's no way anyone can keep their sanity long term.

I mentioned that ts-to-mls should be robust and resilient, but not incorrect in doing so. For now, it's fine to reject this. Maybe we can raise a warning somewhere, and simply ignore the overloads except for the last one (which is usually the most general).

Reply:
Thank you for your advice! I think I can also add some comments after the generated type definitions so that we can find them easily.

Reply:
So far, the output would be like this:
```ts
def XX'f[T]: (T) -> ((anything) -> (string)) /* warning: the overload of function f may be not supported yet */
```
whose form is the most general one

Reply:
Ok, looks good! Except "may not be" -> "is not".

Reply:
```ts
function f<T>(x: T, n: number): string;
function f<U>(x: U, n: boolean): string;
function f<T>(x: T, n) {
    return "";
}
```

Should I output the warning for code like this? Or merely generate the type definition? If no warning should be outputted, guess I need to add some logic to check if two type parameters list are compatible with each other.

Reply:
Of course you should warn here too. Trying to see if overloads can be merged without loss of precision is a hugely complex problem, and it'd be totally useless to work on it since the whole point of overload is to handle situations where they can't be merged.

Reply:
So I just raise a warning when the TypeScript code tries to overload a function with type parameters. Like this:
```ts
def XX'f[T]: (T) -> ((anything) -> (string)) /* warning: the overload of function f is not supported yet. */
```

Reply:
Sure 👍